controllers/installer_controller.go

```diff
@@ -69,37 +123,131 @@ func (r *InstallerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	if len(installer.Spec.URL) == 0 {
 		return reconcile.Result{}, fmt.Errorf("url is not specified")
 	}
+	devtronVersion := strings.ReplaceAll(installer.Spec.URL, "https://raw.githubusercontent.com/devtron-labs/devtron/", "")
+	devtronVersion = strings.ReplaceAll(devtronVersion, "/manifests/installation-script", "")
+	if len(devtronVersion) > 8 {
+		devtronVersion = "v1"
+	}
 	updated := false
+
 	if hasSpecChanged(installer) {
 		fmt.Println("url changed")
 		installer.Status.Sync.Status = installerv1alpha1.SyncStatusCodeOutOfSync
 		installer.Status.Sync.StatusMessage = installerv1alpha1.SyncStatusMessageOutOfSync
 		installer.Status.Sync.URL = installer.Spec.URL
 		installer.Spec.ReSync = false
 		updated = true
+		objectEventType = SpecChanged
 	} else if shouldDownload(installer) {
 		fmt.Println("should download")
 		err := r.sync(installer)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
 		updated = true
+		objectEventType = Downloaded
 	} else if shouldApply(installer) {
 		fmt.Println("applying")
 		r.apply(installer)
 		updated = true
+		objectEventType = Applied
 	}
 
+	//TODO - setup correct event trigger points
 	if updated {
 		fmt.Println("updating")
+		var payload *TelemetryEventDto
+		fromCache := true
+		if installEvent == -1 {
+			fromCache = false
+		}
+		UCID, cm, err := r.getUCID(fromCache)
+		if err != nil {
+			r.Log.Error(err, "failed to get ucid from config map")
+		}
+		if cm != nil && cm.Data != nil && installEvent == -1 {
+			dataMap := cm.Data
+			installEventStr := dataMap["installEvent"]
+			installEvent, err = strconv.Atoi(installEventStr)
+			if err != nil {
+				installEvent = 1 // there might be no key in first time in cm get
+				r.Log.Error(err, "failed to fet install event status from cm")
+			}
+		}
 		err = r.Client.Update(context.Background(), installer)
 		if err != nil {
+			if installEvent == 1 {
+				payload = &TelemetryEventDto{UCID: UCID, Timestamp: time.Now(), EventType: InstallationInterrupt, DevtronVersion: devtronVersion}
+			} else if installEvent == 2 {
+				payload = &TelemetryEventDto{UCID: UCID, Timestamp: time.Now(), EventType: UpgradeInterrupt, DevtronVersion: devtronVersion}
+			}
+			if installEvent == -1 {
+				payload.EventType = InstallationApplicationError
```

Comment:
If operator restarts then installEvent == -1 but cm.Data != nil and cm.Data.installEvent != nil in that case installEvent will remain -1, but this is not InstallationApplicationError, this is upgrade event